### PR TITLE
Add fixed workflow for segmenting layered media.

### PR DESCRIPTION
**Pull request by jwboth**
Segmentation of geoemtries is very useful for images as those of the FluidFlower. The workflow added here will not work for all scenarios, but it has shown to work sufficiently well for the benchmark geometry. Finding the right parameters is not necessarily easy, and tuning will in most cases be required.

### CODE REVIEW

File: src/daria/utils/segmentation.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+from scipy import ndimage as ndi
+
+
+def segment(img: np.ndarray, **kwargs) -> np.ndarray:
+    """
+    Prededfined workflow for segmenting an image based on
+    watershed segmentation. In addition, denoising is used.
+
+    Args:
+        img (np.ndarray): input image in RGB color space
```
Comment (EStorvik) [inline src/daria/utils/segmentation.py:17]: Should the kwargs be mentioned here as well?

**Review by EStorvik — APPROVED** [review]
Looks very good! It is a bit difficult to imagine whether all of the routines function as i guess they should just by looking at it. Are you planning to include some example where it is used as well?

**Comment by jwboth** [comment]
> Looks very good! It is a bit difficult to imagine whether all of the routines function as i guess they should just by looking at it. Are you planning to include some example where it is used as well?

I completely agree. It is difficult to assess how this works just by looking at the source code. I will try to add a simple example. This should work for coarse images in fact, so there is no excuse.

**Comment by jwboth** [comment]
> Looks very good! It is a bit difficult to imagine whether all of the routines function as i guess they should just by looking at it. Are you planning to include some example where it is used as well?

New update which I wanted to add before merging: I replaced the old example now. Could you please have a quick look @EStorvik ?

File: src/daria/utils/segmentation.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+        np.ndarray: updated labeled image
+    """
+    # Top
+    labels[:thickness, :] = labels[thickness : thickness + 1, :]
+    # Bottom
+    labels[-thickness - 1 : -1, :] = labels[-thickness - 1 : -thickness, :]
+    # Left
+    labels[:, :thickness] = labels[:, thickness : thickness + 1]
+    # Right
+    labels[:, -thickness - 1 : -1] = labels[:, -thickness - 1 : -thickness]
```
Comment (jwboth) [inline src/daria/utils/segmentation.py:228]:
```suggestion
    labels[:, -thickness :] = labels[:, -thickness - 1 : -thickness]
```
Also cover the boundary column.

File: src/daria/utils/segmentation.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+        labels (np.ndarray): labeled image
+        thickness (int): thickness of boundary which should be overwritten
+
+    Returns:
+        np.ndarray: updated labeled image
+    """
+    # Top
+    labels[:thickness, :] = labels[thickness : thickness + 1, :]
+    # Bottom
+    labels[-thickness - 1 : -1, :] = labels[-thickness - 1 : -thickness, :]
```
Comment (jwboth) [inline src/daria/utils/segmentation.py:224]:
```suggestion
    labels[-thickness :, :] = labels[-thickness - 1 : -thickness, :]
```
Include the last row.

**Review by EStorvik — APPROVED** [review]
I think it looks very good! The example is cool as well!

File: examples/segmentation.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    "rescaling factor": 0.3,
+    "markers disk radius": 10,
+    "threshold": 20,
+    "gradient disk radius": 2,
+}
+
+# Determine labels. To better understand the choice of the tuning parameters (and in
+# practical situations for finding choices for these) set verboisty to True. Essentially
+# relevant quantities are plotted.
+labels = daria.segment(image, verbosity=False, **config)
```
[the remark this answers is not in the document]
Comment (jwboth) [reply examples/segmentation.py:36]: Agree. I will do that. Due to current conflicts (git) in my working directory, I will however postpone this and merge when this is finished.

[the remark this answers is not in the document]
Comment (EStorvik) [reply examples/segmentation.py:36]: Great!

[the remark this answers is not in the document]
Comment (jwboth) [reply examples/segmentation.py]: Done.

**Review by EStorvik — APPROVED** [review]
I think it looks good, and the supervised version is also cool. However, are you satisfied with the result from the supervised part in the example? It seems like there are some strange things happening on the middle right side of the image. I have no clue though, and it might be that this is precisely the correct geometry.

**Comment by jwboth** [comment]
> I think it looks good, and the supervised version is also cool. However, are you satisfied with the result from the supervised part in the example? It seems like there are some strange things happening on the middle right side of the image. I have no clue though, and it might be that this is precisely the correct geometry.

Yes and no. Obviously it is not perfect. But this is mainly due to the input image. Even with the bare eye in some parts of the image it is almost impossible to predict a boundary between layers. And this transfers to the image based segmentation. The right thing to do here, would be to continue the segmentation (something I am currently also working on) which grabs all segments which are not chosen satisfactorily. A further analysis with contrast enhancement would have to be applied to have a chance. For now, I am happy with the current setup, as the most important to me was the detection of the more obvious layers (e.g. the lighter one just below the top ESF layer) which also results in sharp discontinuities in the concentration signal. To correct for it, a better segmentation is needed. Hopefully, the current approach is heading towards the right direction.

I am merging.

**Pull request merged**